### PR TITLE
fix(fast_dispatch):  saving input tensor using ctx.save_for_backward

### LINKED ISSUE #237: Potential Memory Leak in GatingEncoder/Decoder of Fast_Dispatch
HI, during my usage of the tutel fast_dispatch, I noticed that when `gradient accumulation` is enabled, the GPU memory consumption significantly exceeds that when `gradient accumulation` is OFF, and the consumed memory increases with the growth of `gradient accumulation` steps. 

Upon reviewing the relevant code, I found that in the [`Encode`/`Decode` implementation](https://github.com/microsoft/tutel/blob/main/tutel/impls/fast_dispatch.py#L19), the input tensor is directly placed in the `ctx` without invoking `save_for_backward`. 
According to the [PyTorch documentation](https://pytorch.org/docs/stable/generated/torch.autograd.function.FunctionCtx.save_for_backward.html), this practice may lead to a memory leak. 

I attempted to fix this issue, and the memory consumption no longer grows with the number of gradient accumulation steps.

### PR DESCRIPTION
**Pull request by KimmiShi**
fix #237

### CODE REVIEW

**Comment by ghostplant** [comment]
LGTM. Thanks!
